nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceProvider.java

```diff
@@ -253,38 +252,44 @@ public CompletableFuture<Void> enableControllerService(final ControllerServiceNo
+                logger.warn("Failed to enable service {} because it is not currently valid", controllerServiceNode);
+            } catch (Exception e) {
+                logger.error("Failed to enable {}", controllerServiceNode, e);
+                if (this.bulletinRepo != null) {
+                    this.bulletinRepo.addBulletin(BulletinFactory.createBulletin("Controller Service",
+                            Severity.ERROR.name(), "Could not start " + controllerServiceNode + " due to " + e));
                 }
             }
         }
+    }
 
-        if (shouldStart) {
-            for (ControllerServiceNode controllerServiceNode : serviceNodes) {
-                try {
-                    final Future<Void> future = enableControllerServiceAndDependencies(controllerServiceNode);
-
-                    future.get(30, TimeUnit.SECONDS);
-                    logger.debug("Successfully enabled {}; service state = {}", controllerServiceNode, controllerServiceNode.getState());
-                } catch (final ControllerServiceNotValidException csnve) {
-                    logger.warn("Failed to enable service {} because it is not currently valid", controllerServiceNode);
-                } catch (Exception e) {
-                    logger.error("Failed to enable {}", controllerServiceNode, e);
-                    if (this.bulletinRepo != null) {
-                        this.bulletinRepo.addBulletin(BulletinFactory.createBulletin("Controller Service",
-                                Severity.ERROR.name(), "Could not start " + controllerServiceNode + " due to " + e));
-                    }
+    private Collection<ControllerServiceNode> filterStartableControllerServices(final Collection<ControllerServiceNode> serviceNodes) {
+        Collection<ControllerServiceNode> startableServiceNodes = new HashSet<>();
+        for (ControllerServiceNode serviceNode : serviceNodes) {
+            boolean isStartable = true;
+            List<ControllerServiceNode> requiredServices = serviceNode.getRequiredControllerServices();
+            for (ControllerServiceNode requiredService : requiredServices) {
+                if (!requiredService.isActive() && !serviceNodes.contains(requiredService)) {
+                    isStartable = false;
+                    logger.error("Will not start {} because its required service {} is not active and is not part of the collection of things to start", serviceNode, requiredService);
                 }
             }
+            if (isStartable) {
+                startableServiceNodes.add(serviceNode);
+            }
+        }
+        if (startableServiceNodes.size() < serviceNodes.size()) {
+            // If any service was removed, then recheck all remaining services because the removed one might be required by another service.
+            startableServiceNodes = filterStartableControllerServices(startableServiceNodes);
         }
+        return startableServiceNodes;
```

Comment:
This approach to build a `HashSet` for every invocation and recursively call this method seems like it could get expensive for flows with larger numbers of Controller Services.

Reply:
@exceptionfactory Thanks for reviewing this and giving feedback.

This method being changed: 
StandardControllerServiceProvider.enableControllerServices(
        final Collection<ControllerServiceNode> **serviceNodesIn**)

In the existing code and in the new code, the controller services in the **serviceNodes** parameter
contains only controller services which are supposed to be enabled. 
The caller decides which services it intends to enable.
If a controller service is not supposed to be started on Nifi restart (i.e. scheduledState is disabled), then it should not be (and is not) put in **serviceNodes** by the existing calling code.

To recap, the method I am changing receives a list of only the services that the caller wants to enable (based on the caller's decision logic). It does **not**  get a list of all controller services when NiFi is being restarted.

Reply:
So this change does **not** enable any controller services which the caller of this method did **not** ask to enable. And does not introduce any new side effect of enabled unwanted controller services.

Reply:
@exceptionfactory I refactored the code based on your comments on its structure.